linera-views/src/hashable_wrapper.rs

```diff
@@ -29,26 +33,36 @@ enum KeyTag {
     Hash,
 }
 
+/// That is used for accessing whether we need to delete the storage first before
+/// writing
+pub(crate) trait DeleteStorageFirst {
+    fn delete_storage_first(&self) -> bool;
+}
```

Comment:
We could do such a thing. However, that would potentially increase the size of the memory used.

Reply:
I'd expect the enum and the option type to have a similar memory layout.

Reply:
You are likely right and I still think it is not a great idea. You are relying on the compiler to avoid the unavoidable fact that speed requires you to have leaky abstractions. Because with the choice you propose, we would have the information of being cleared in two places: In `Cleared` and in the `delete_storage_first`. This is not pretty.

Reply:
Using another's component state (not unlike [intrusive pointers](https://www.boost.org/doc/libs/1_38_0/libs/smart_ptr/intrusive_ptr.html) in C++/boost) feels like a hack. Specifically, the code is longer (since O(N) supported components need to implement the trait) and less readable (since correctness invariants are not locally enforceable any more).
That would be ok if there was a compelling reason but I don't see it.

Reply:
Depending on the compiler optimization in order to be sure that the size of an enum is the same as the one of an option is surely a hack if you choose to use that word. In release mode I agree with you that it would pass, in debug, I am not so sure.

The argument about invariants goes both ways: The `delete_storage_first` is set to true when `clear` is called. When using the `enum` construction, we would have the clear being set in two places and the consistency of both assignations is certainly an invariant which would not be checkable since both parts would ignore what the other is doing.

The code as it is implemented simply reflects the functionality as they are. We do not want a `clear+flush` to leave anything on the storage and that imposes some constraints. This design did not come easily. The first design used the `delete` (which might be the most economical though rough approach to the question if deleting on the storage), the second design used relatively complex designs in particular for `LogView`, `QueueView`, `RegisterView` and `KeyValueStoreView`. Then as a result of the reviewing process, this was simplified and somewhat unexpectedly we ended up with the same approach on all views. And so now this becomes a trait.

Your mention of `intrusive_ptr` is relevant inasmuch as yes if a pointer has its own mechanics for reference counting there is no point in duplicating it. And yes I would agree on that. And yes I would use an `intrusive_ptr` if needed.

Reply:
The memory layout of Rust enums can be controlled in a documented way if needed: https://doc.rust-lang.org/reference/type-layout.html#primitive-representation-of-enums-with-fields

I mentioned intrusive pointers because I thought the word "intrusive" was relevant to the discussion. I'm sure they have their use cases.

Reply:
Ok, for that question of the size of the enum I admit that I found the documentation to be unclear and instead I ran experience: ```
type Cont1<C> = Option<C>;

struct Cont2<C> {
    opt: Option<C>,
}

#[repr(u8)]
enum Cont3<C> {
    Cleared,
    Unknown,
    Known(C),
}

fn main() {
    type C = Vec<u8>;
    println!("|Cont1|={}", std::mem::size_of::<Cont1<C>>());
    println!("|Cont2|={}", std::mem::size_of::<Cont2<C>>());
    println!("|Cont3|={}", std::mem::size_of::<Cont3<C>>());
}````

with the result being ```
|Cont1|=24
|Cont2|=24
|Cont3|=32````

Reply:
Interestingly (and ironically), I'm finding the size 32 only for the enum when `repr(u8)` is added. For some reason, the padding caused by alignment is larger in this case.

Anyways, we're splitting hairs a little bit here. 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a8943cac57ce60df402e3065aa4e04f3

Reply:
I did test the `#[repr(u8)]` question and found that with and without the size was `32`.
Further investigation showed that the change from 32 to 24 (on my laptop which is x86_64) occurred from version 1.75.0 to 1.76.0 released on the 8th of February 2024. What I take from that discussion is:
* The code `memdump` for getting the binary representation of an object, that could be useful for further works.
* The Rust Compiler can change its alignment without warning (there is a mention of changes to the struct in the release note of 1.76.0, but not for the enum).
* On the other hand, if we use the `#[repr(u8)]` or similar, we can force the rust compiler to a fixed alignment. This is useful for compatibility of pointers with other languages. So, that alignment would not change from one version to the next.
* In particular with `#[repr(C)]` we have a size of 32 in version 1.75.0 and 1.76.0 since the C-ABI does not change.
* It is difficult to use the manual for such purposes. Much of this is not well documented and one has to experiment.

I initially wrote "You are relying on the compiler to avoid the unavoidable fact that speed requires you to have leaky abstractions" + "This is not pretty". I think that these two assessments have been born out by subsequent investigation.

I cannot see any other aspect to discuss in that question. What is your decision about that PR?

Reply:
Let's give some time for @afck to chime in.

It's normal that adding unexpected public APIs triggers a discussion. Technically, the new trait makes the combinator slightly less general than anticipated -- unless views always implement DeleteStorageFirst in practice but then should it part of the View trait? (but then again do we need this?)

While I enjoyed looking into the memory layout, I don't think small differences caused by memory alignment generally matter for us. (They may also disappear when the value is inserted into a larger object)

Reply:
Ok.

Here is my argumentation about the design of the trait:
* It did happened in previous work that the `delete_storage_first` was introduced in all the views and that the code became the same. It was unexpected to me.
* So, the introduction of a trait definition is a natural conclusion.
* The function `delete_storage_first()` is not included into the `View` because it would expose to the users some internal function. The trait `DeleteStorageFirst` is not a public trait, it is a `pub(crate)`.

The issue of the size is precisely where we have to be careful. We need to reduce memory usage as much as possible.

Reply:
I'm still hoping we need neither the duplication nor the trait: https://github.com/linera-io/linera-protocol/pull/1779#discussion_r1528311012
(But I might be misunderstanding something.)